### PR TITLE
Add support for gpios

### LINKED ISSUE #69: Scaled joint trajectory controller: `rt_active_goal_` is not declared
On develop branch I get a bunch of build errors related to `rt_active_goal_`. It looks like it is not declared anywhere.

Some of the errors are:

```
scaled_joint_trajectory_controller.cpp:153:11: error: could not convert ‘((ur_controllers::ScaledJointTrajectoryController*)this)->ur_controllers::ScaledJointTrajectoryController::<anonymous>.joint_trajectory_controller::JointTrajectoryController::rt_active_goal_’ from ‘joint_trajectory_controller::JointTrajectoryController::RealtimeGoalHandleBuffer’ {aka ‘realtime_tools::RealtimeBuffer<std::shared_ptr<realtime_tools::RealtimeServerGoalHandle<control_msgs::action::FollowJointTrajectory> > >’} to ‘bool’
  153 |       if (rt_active_goal_)
```

```
scaled_joint_trajectory_controller.cpp:181:27: error: ‘using RealtimeGoalHandleBuffer = class realtime_tools::RealtimeBuffer<std::shared_ptr<realtime_tools::RealtimeServerGoalHandle<control_msgs::action::FollowJointTrajectory> > >’ {aka ‘class realtime_tools::RealtimeBuffer<std::shared_ptr<realtime_tools::RealtimeServerGoalHandle<control_msgs::action::FollowJointTrajectory> > >’} has no member named ‘reset’
  181 |           rt_active_goal_.reset();
```

### PR DESCRIPTION
**Pull request by destogl**
Here a bit larger PR but it should work. There is a lot of refactoring happening to also have a manual.

@AndyZe before reviewing, please check #58 and merge, if works for you. Then this PR can be rebased and gets smaller.

@MGBla, @Abishalini could you please look into this on hardware. Please see the example command on how to start this. You should understand all, if not, then please tell me what is not clear in the manual.

Install the package and dependencies in a clean ros2 workspace. The following commands assume that you are in the root of your workspace and `src` folder exists.
```
git clone git@github.com:PickNikRobotics/Universal_Robots_ROS2_Driver.git src/Universal_Robots_ROS2_Driver --branch add-support-for-gpios
vcs import src --skip-existing --input src/Universal_Robots_ROS2_Driver/Universal_Robots_ROS2_Driver.repos
rosdep install --ignore-src --from-paths src -y -r
colcon build --cmake-args -DCMAKE_BUILD_TYPE=Release
source install/setup.bash
```

The following should be tested:

1. Starting the robot with default values, i.e. (joint_trajectory_controller). Other controllers should work out-of-the-box.
2. Test with starting other controllers:
    - scaled_joint_trajectory_controller
    - forward_position_controller
3. Test movement using two scripts (works for joint_trajectory_controller and forward_position_controller).
4. Test GPIO part by setting/reading values. For setting a value the easiest is to use `rqt`-service caller. There are topics and services with obvious name :)
5. Test MoveIt launch file and trajectory planing and execution.

Should replace #42. Closes #69

### CODE REVIEW

File: ur_bringup/launch/ur3.launch.py [hunk 1]
```diff
@@ -26,7 +26,6 @@ def generate_launch_description():
     declared_arguments.append(
         DeclareLaunchArgument(
             "robot_ip",
-            default_value="",
```
Comment (AndyZe) [inline ur_bringup/launch/ur3.launch.py:29]: Nit: this type of change would be better in a separate PR

File: ur_controllers/src/gpio_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+  for (size_t i = 0; i < 2; ++i)
+  {
+    io_msg_.analog_in_states[i].state =
+        static_cast<float>(state_interfaces_[i + StateInterfaces::ANALOG_INPUTS].get_value());
+    io_msg_.analog_in_states[i].domain =
+        static_cast<uint8_t>(state_interfaces_[i + StateInterfaces::ANALOG_IO_TYPES].get_value());
+  }
+
+  for (size_t i = 0; i < 2; ++i)
+  {
```
Comment (AndyZe) [inline ur_controllers/src/gpio_controller.cpp:182]: Combine with prevous for loop, since the range is the same

File: ur_controllers/src/gpio_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    // io async success
+    command_interfaces_[CommandInterfaces::IO_ASYNC_SUCCESS].set_value(2.0);
+    command_interfaces_[req->pin].set_value(static_cast<double>(req->state));
+
+    RCLCPP_INFO(node_->get_logger(), "Setting digital output '%d' to state: '%1.0f'.", req->pin, req->state);
+
+    while (command_interfaces_[CommandInterfaces::IO_ASYNC_SUCCESS].get_value() != 2.0)
+    {
+      // setting the value is not yet finished
```
Comment (AndyZe) [inline ur_controllers/src/gpio_controller.cpp:256]: Add TODO so we can find this easily in the future

File: ur_controllers/src/gpio_controller.cpp [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+  {
+    // io async success
+    command_interfaces_[CommandInterfaces::IO_ASYNC_SUCCESS].set_value(2.0);
+    command_interfaces_[CommandInterfaces::ANALOG_OUTPUTS_CMD + req->pin].set_value(static_cast<double>(req->state));
+
+    RCLCPP_INFO(node_->get_logger(), "Setting analog output '%d' to state: '%1.0f'.", req->pin, req->state);
+
+    while (command_interfaces_[CommandInterfaces::IO_ASYNC_SUCCESS].get_value() != 2.0)
+    {
+      // setting the value is not yet finished
```
Comment (AndyZe) [inline ur_controllers/src/gpio_controller.cpp:273]: Add TODO

File: ur_controllers/src/gpio_controller.cpp [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+      std::this_thread::sleep_for(std::chrono::milliseconds(50));
+    }
+
+    resp->success = static_cast<bool>(command_interfaces_[20].get_value());
+    return resp->success;
+  }
+  else if (req->fun == req->FUN_SET_ANALOG_OUT && req->pin >= 0 && req->pin <= 2)
+  {
+    // io async success
+    command_interfaces_[CommandInterfaces::IO_ASYNC_SUCCESS].set_value(2.0);
```
Comment (AndyZe) [inline ur_controllers/src/gpio_controller.cpp:266]: Why is this value of 2.0 significant? It prob should be an enum

File: ur_controllers/src/gpio_controller.cpp [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    command_interfaces_[CommandInterfaces::ANALOG_OUTPUTS_CMD + req->pin].set_value(static_cast<double>(req->state));
+
+    RCLCPP_INFO(node_->get_logger(), "Setting analog output '%d' to state: '%1.0f'.", req->pin, req->state);
+
+    while (command_interfaces_[CommandInterfaces::IO_ASYNC_SUCCESS].get_value() != 2.0)
```
Comment (AndyZe) [inline ur_controllers/src/gpio_controller.cpp:271]: Use enum for the 2.0 here, too

File: ur_controllers/src/gpio_controller.cpp [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+  {
+    // reset success flag
+    command_interfaces_[CommandInterfaces::SCALING_ASYNC_SUCCESS].set_value(2.0);
+    // set commanding value for speed slider
+    command_interfaces_[CommandInterfaces::SPEED_SCALING_CMD].set_value(
+        static_cast<double>(req->speed_slider_fraction));
+
+    while (command_interfaces_[CommandInterfaces::SCALING_ASYNC_SUCCESS].get_value() != 2.0)
+    {
+      // setting the value is not yet finished
```
Comment (AndyZe) [inline ur_controllers/src/gpio_controller.cpp:300]: Add TODO

File: ur_description/urdf/ur.ros2_control.xacro [hunk 8]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0"?>
 <robot xmlns:xacro="http://www.ros.org/wiki/xacro">
 
-  <xacro:macro name="ur5_system_position_only" params="name prefix
+  <xacro:macro name="ur_ros2_control" params="name prefix
     use_fake_hardware:=^|false fake_sensor_commands:=^|false
```
Comment (AndyZe) [inline ur_description/urdf/ur.ros2_control.xacro:5]: delete ^|

File: ur_description/urdf/ur.ros2_control.xacro [hunk 9]
```diff
@@ -126,6 +126,120 @@
         <state_interface name="torque.y"/>
         <state_interface name="torque.z"/>
       </sensor>
+
+      <!-- NOTE The following are joints used only for testing with fake hardware and will chnage int the future -->
```
Comment (AndyZe) [inline ur_description/urdf/ur.ros2_control.xacro:130]: change in

File: ur_robot_driver/launch/ur5_bringup.launch [hunk 10]
```diff
@@ -5,7 +5,7 @@
   <arg name="reverse_port" default="50001" doc="Port that will be opened by the driver to allow direct communication between the driver and the robot controller."/>
   <arg name="script_sender_port" default="50002" doc="The driver will offer an interface to receive the program's URScript on this port. If the robot cannot connect to this port, `External Control` will stop immediately."/>
   <arg name="tf_prefix" default="" doc="tf_prefix used for the robot."/>
-  <arg name="controllers" default="joint_state_controller scaled_pos_joint_traj_controller speed_scaling_state_controller force_torque_sensor_controller" doc="Controllers that are activated by default."/>
+  <arg name="controllers" default="joint_state_broadcaster scaled_pos_joint_traj_controller speed_scaling_state_controller force_torque_sensor_controller" doc="Controllers that are activated by default."/>
```
Comment (AndyZe) [inline ur_robot_driver/launch/ur5_bringup.launch:8]: I vote to delete some of these args from urX_bringup.launch and let the defaults in ur_common.launch take care of them

File: ur_robot_driver/src/hardware_interface.cpp [hunk 11]
```diff
@@ -477,4 +575,95 @@ void URPositionHardwareInterface::handleRobotProgramState(bool program_running)
...
+    }
+  }
+
+  for (size_t i = 0; i < 2; ++i)
+  {
+    if (standard_analog_output_cmd_[i] != standard_analog_output_[i])
+    {
+      io_async_success_ = ur_driver_->getRTDEWriter().sendStandardAnalogOutput(i, standard_analog_output_cmd_[i]);
+
+      if (io_async_success_ != 2.0)
```
Comment (AndyZe) [inline ur_robot_driver/src/hardware_interface.cpp:625]: Replace with enum

File: ur_robot_driver/launch/ur5_bringup.launch (hunk 10 above)
Comment (AndyZe) [reply ur_robot_driver/launch/ur5_bringup.launch:8]: ^ @destogl and @gavanderhoorn I would appreciate comments here.

Currently many of the args are passed like this:

urX_bringup.launch --> ur_common.launch --> ur_control.launch

I think we should define the defaults in ur_control.launch and delete the redundant definitions in the other 2 files. Less mindless passing of args is a good thing.

File: ur_controllers/src/gpio_controller.cpp (hunk 5 above)
Comment (livanov93) [reply ur_controllers/src/gpio_controller.cpp:266]: It represents success. Enum would be better for sure...

File: README.md [hunk 12]
```diff
@@ -2,96 +2,143 @@
...
 
-Start the `joint_state_controller`:
+For starting the driver there are three main launch files in the `ur_bringup` package.
 
-```
-ros2 control load_start_controller joint_state_controller
-```
+  - `ur_control.launch.py` - starts ros2_control node including hardware interface, joint state broadcaster and a controller.
+  - `ur_moveit.launch.py` - start everything from `ur_control.launch.py` plus an example scenario with [MoveIt2](https://moveit.ros.org/).
+  - `ur_dashboard_client.launch.py` - start dashboard client for UR robots.
```
[the remark this answers is not in the document]
Comment (livanov93) [reply README.md]: Which one?

[the remark this answers is not in the document]
Comment (AndyZe) [reply README.md]: ur_dashboard_client does not need to be launched, I thought

File: ur_robot_driver/launch/ur5_bringup.launch (hunk 10 above)
Comment (AndyZe) [reply ur_robot_driver/launch/ur5_bringup.launch]: I went ahead and pushed a commit to only define each arg default in one place. We can revert if you feel strongly

File: ur_controllers/src/gpio_controller.cpp (hunk 5 above)
Comment (AndyZe) [reply ur_controllers/src/gpio_controller.cpp:266]: @livanov93 can you please take a look? It seems like 2.0 is a typo.

In gpio_controller.hpp, `20` is IO_ASYNC_SUCCESS. Not `2.0`

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (destogl) [reply README.md]: I am not sure for the hardware. So it is optional.

File: ur_bringup/launch/ur3.launch.py (hunk 1 above)
Comment (destogl) [reply ur_bringup/launch/ur3.launch.py:29]: I would keep those here. It doesn't make much sense to remove them now. But yeah I agree, this PR become huge... It's just that many changes were needed and time is limited :)

File: ur_robot_driver/launch/ur5_bringup.launch (hunk 10 above)
Comment (destogl) [reply ur_robot_driver/launch/ur5_bringup.launch]: `ur_common.launch` and `ur_control.launch` are not use at all... These are the leftovers from ROS1. Didn't want to delete those in this PR since it is huge...

Oh, this line change happened by find-replace all. And I wasn't careful when commiting.

File: ur_robot_driver/src/hardware_interface.cpp (hunk 11 above)
Comment (destogl) [reply ur_robot_driver/src/hardware_interface.cpp:625]: This is OK. There is no enum for this. Let's keep this as it is for the time being. Otherwise, will get tricky with dependencies between packages.

**Comment by destogl** [comment]
@AndyZe I added here also MoveIt launch file to make it easy for people to test. If this goes well, we are almost there. Then only cosmetic changes are left.

**Comment by MGBla** [comment]
Most of the pins seem to work except i cannot set the analog_out[1] current. Another problem i encountered is that after setting an input using the service call i wasn't able to change it in the teach panel. I think as the safety device the teach panel should always be able to override values we set using service calls.

**Comment by Abishalini** [comment]
I had the chance to test today and here are the things we noticed -
* We need to load and start the controllers manually 
```
ros2 control load_start_controller joint_state_broadcaster
ros2 control load_start_controller joint_trajectory_controller
```
* Using MoveIt does not work (something with planning scene initialization failure)
* We could set the analog_out[1] current. I didn't have the chance to check if the teach pendant can set IO values after using the service call

File: ur_moveit_config/config/kinematics.yaml [hunk 13]
```diff
@@ -1,4 +1,4 @@
-ur_manipulator:
+ur5e_manipulator:
```
Comment (destogl) [inline ur_moveit_config/config/kinematics.yaml:1]: I don't think this change is sensible... This brakes all other UR models.

File: ur_bringup/launch/ur_moveit.launch.py [hunk 14]
```diff
@@ -406,7 +406,7 @@ def generate_launch_description():
         executable="rviz2",
         name="rviz2",
         output="log",
-        arguments=["-d", rviz_config_file],
+        arguments=["-d", ""],
```
Comment (destogl) [inline ur_bringup/launch/ur_moveit.launch.py:409]: Why shouldn't we provide default rviz config?

Comment (AndyZe) [reply ur_bringup/launch/ur_moveit.launch.py:409]: I don't expect it to get merged like this but I find loading the RViz config causes a simulation failure somehow. Probably timing-related. We can revert this commit.

File: ur_moveit_config/config/kinematics.yaml (hunk 13 above)
Comment (AndyZe) [reply ur_moveit_config/config/kinematics.yaml:1]: Yep, I don't expect it to get merged like this but it is some last-minute debugging.

Comment (destogl) [reply ur_moveit_config/config/kinematics.yaml]: I corrected this now

File: ur_controllers/src/gpio_controller.cpp [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+  {
+    resp->success = false;
+    return false;
+  }
+}
+
+bool GPIOController::setSpeedSlider(ur_msgs::srv::SetSpeedSliderFraction::Request::SharedPtr req,
+                                    ur_msgs::srv::SetSpeedSliderFraction::Response::SharedPtr resp)
+{
+  if (req->speed_slider_fraction >= 0.01 && req->speed_slider_fraction <= 1.0)
```
[the remark this answers is not in the document]
Comment (destogl) [reply ur_controllers/src/gpio_controller.cpp:290]: I think it is safe to assume that users give as values in percentage... So I will just divide the input with 100

[the remark this answers is not in the document]
Comment (gavanderhoorn) [reply ur_controllers/src/gpio_controller.cpp]: That would be a violation of the [service contract](https://github.com/ros-industrial/ur_msgs/blob/cec984f35cce1ba8435318a585530cbea0a8fa8c/srv/SetSpeedSliderFraction.srv#L1-L4).

File: ur_moveit_config/config/kinematics.yaml (hunk 13 above)
Comment (AndyZe) [reply ur_moveit_config/config/kinematics.yaml]: FYI this solved this error:

`Error:   Semantic description is not specified for the same robot as the URDF`

**Comment by AndyZe** [comment]
This is now working great for me in simulation!  :+1:

**Comment by AndyZe** [comment]
I"ll make issues for what Abi and Marvin have mentioned:

- Cannot set analog_out[1] current
- after setting an input using the service call i wasn't able to change it in the teach panel. I think as the safety device the teach panel should always be able to override values we set using service calls.

**Comment by gavanderhoorn** [comment]
Should the divergence from the ROS 1 version of the driver wrt speed slider service be tracked in an issue?

**Comment by AndyZe** [comment]
^ yes, I think so
